staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,417 @@
+	"reflect"
+	"sync"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+func intervalFromEvents(events []*watchCacheEvent) *watchCacheInterval {
+	startIndex, endIndex, locker := 0, len(events), &sync.Mutex{}
+	indexer := func(i int) *watchCacheEvent {
+		if len(events) == 0 {
+			return nil
+		}
+		return events[i]
+	}
+	indexValidator := func(_ int) bool { return true }
+	wci := newCacheInterval(startIndex, endIndex, indexer, indexValidator, locker)
+
+	return wci
+}
+
+func bufferFromEvents(events []*watchCacheEvent) *watchCacheIntervalBuffer {
+	wcib := &watchCacheIntervalBuffer{}
+	wcib.buffer = make([]*watchCacheEvent, bufferSize)
+	copy(wcib.buffer, events)
+	wcib.endIndex = len(events)
+	return wcib
+}
+
+func generateEvents(start, end int) []*watchCacheEvent {
+	n := end - start
+	events := make([]*watchCacheEvent, n)
+	for i := 0; i < n; i++ {
+		events[i] = &watchCacheEvent{
+			Type: watch.Added, Object: makeTestPod(
+				fmt.Sprintf("pod%d", start+i+1),
+				uint64(start+i+1),
+			),
+		}
+	}
+	return events
+}
+
+func TestIntervalBufferIsFull(t *testing.T) {
+	cases := []struct {
+		endIndex int
+		expected bool
+	}{
+		{endIndex: bufferSize - 1, expected: false},
+		{endIndex: bufferSize, expected: true},
+		{endIndex: bufferSize + 1, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{endIndex: c.endIndex}
+		actual := wcib.isFull()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferIsEmpty(t *testing.T) {
+	cases := []struct {
+		startIndex int
+		endIndex   int
+		expected   bool
+	}{
+		{startIndex: 0, endIndex: 10, expected: false},
+		{startIndex: 5, endIndex: 20, expected: false},
+		{startIndex: 50, endIndex: 50, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{
+			startIndex: c.startIndex,
+			endIndex:   c.endIndex,
+		}
+		actual := wcib.isEmpty()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferNext(t *testing.T) {
+	cases := []struct {
+		events []*watchCacheEvent
+	}{
+		{
+			events: []*watchCacheEvent{
+				{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+				{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+				{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+			},
+		},
+		{
+			events: []*watchCacheEvent{},
+		},
+	}
+
+	for _, c := range cases {
+		wcib := bufferFromEvents(c.events)
+
+		received := 0
+		for {
+			event, ok := wcib.next()
+			if !ok {
+				if received != len(c.events) {
+					t.Errorf("unexpected bool value indicating empty buffer")
+				}
+				break
+			}
+			if received == len(c.events) {
+				t.Errorf("unexpected bool value indicating non-empty buffer")
+				break
+			}
+			if event == nil {
```

Comment:
This if isn't needed - this will fail the deep-equal call below anyway...

Reply:
Same in below tests.

Reply:
👋🏼 
I mostly kept this because I wanted to log the dereferenced value of the event so that debugging is easier, and not the pointer itself.